examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+import os
+from collections.abc import Mapping
+from functools import partial
+from typing import Any, cast, List, Literal, Optional, Tuple, Union
+
+import torch
+from lightning_utilities.core import is_overridden
+from lightning_utilities.core.apply_func import apply_to_collection
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+from lightning.fabric.fabric import (
+    _PLUGIN_INPUT,
+    _PRECISION_INPUT,
+    _unwrap_objects,
+    Accelerator,
+    Fabric,
+    Logger,
+    Strategy,
+)
+from lightning.fabric.strategies.fsdp import FSDPStrategy
+from lightning.fabric.utilities.rank_zero import rank_zero_warn
+from lightning.fabric.utilities.types import LRScheduler, Optimizable
+from lightning.pytorch.core.module import LightningModule
+
+
+class Trainer:
+    def __init__(
+        self,
+        accelerator: Union[str, Accelerator] = "auto",
+        strategy: Union[str, Strategy] = "auto",
+        devices: Union[List[int], str, int] = "auto",
+        precision: _PRECISION_INPUT = "32-true",
+        plugins: Optional[Union[_PLUGIN_INPUT, List[_PLUGIN_INPUT]]] = None,
+        callbacks: Optional[Union[List[Any], Any]] = None,
+        loggers: Optional[Union[Logger, List[Logger]]] = None,
+        max_epochs: Optional[int] = 1000,
+        max_steps: Optional[int] = None,
```

Comment:
We can consider going iteration-based only. Open for discussion. 
My argument is that it's easier to read, but there are arguments against that would say it's harder to read. So it's probably personal preference.

Reply:
I think that while the loop in iteration-based is easier to follow, the dataloader-handling (with resetting etc.) is harder to read. Also, the epoch-based approach is what people are familiar with from almost all examples of PyTorch and other resources as it (sadly) still is the most common standard.